File: leetcode/python/0876. Middle of the Linked List.py
# Time complexity: O(N), where N is the number of nodes
# Space complexity: O(1)

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def middleNode(self, head):
        temp = head
        length = 0
        while temp != None:
            length += 1
            temp = temp.next
        
        mid = length / 2
        toLoop = length - mid - 1
        if length % 2 == 0:
            toLoop += 1

        while toLoop > 0:
            head = head.next
            toLoop -= 1
        return head

File: leetcode/java/0876. Middle of the Linked List.java
package leetcode.java;

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Time complexity: O(N), where N is the number of nodes
// Space complexity: O(1)

class Solution {
    public ListNode middleNode(ListNode head) {
        int length = 0;
        ListNode temp = head;

        while (temp != null) {
            temp = temp.next;
            length += 1;
        }

        length /= 2;

        while (length > 0) {
            head = head.next;
            length -= 1;
        }

        return head;
    }
}